### PR TITLE
Add custom nth_back for Zip

### PR DESCRIPTION
**Pull request by mdnight**
Implementation of nth_back for Zip.
Part of #54054
r? @scottmcm

### CODE REVIEW

**Comment by timvermeulen** [comment]
Since the underlying iterators may have side-effects, I don't think we can return early when `min_len < n`.

**Comment by koalatux** [comment]
@timvermeulen, returning early might be ok, especially if you are leveraging the underlaying nth_back. The forward next also does short-circuiting, its documentation sais: "If the first iterator returns None, zip will short-circuit and next will not be called on the second iterator." Maybe we can use a similar behaviour here.

**Comment by timvermeulen** [comment]
@koalatux That guarantee is specific to `Zip`, whereas custom implementations of `nth` and `nth_back` are required by the `Iterator` trait to have the same behavior in regards to side effects as the default implementations (i.e. repeated calls to `next`/`next_back`).

So short-circuiting `Zip::nth_back` by forwarding to `nth_back` on the inner iterators is indeed fine, if done carefully, but short-circuiting based on the simple check `min_len < n` is not.

**Comment by Dylan-DPC** [comment]
ping from triage @scottmcm waiting for your review on this

File: src/libcore/iter/adapters/zip.rs [hunk 1]
```diff
@@ -62,6 +62,17 @@ impl<A, B> DoubleEndedIterator for Zip<A, B> where
     fn next_back(&mut self) -> Option<(A::Item, B::Item)> {
         ZipImpl::next_back(self)
     }
+
+    #[inline]
+    fn nth_back(&mut self, mut n: usize) -> Option<(A::Item, B::Item)> {
+        while let Some(x) = ZipImpl::next_back(self) {
```
Comment (scottmcm) [inline src/libcore/iter/adapters/zip.rs]: The default implementation of `nth_back` calls `Iterator::next_back` in a loop, which seems like it'd be just as good as this one?  It looks like `ZipImpl::nth` exists, so I'd have expected the change here to add `ZipImpl::nth_back` and specialize it the same way.

**Review by scottmcm — CHANGES_REQUESTED** [review]

**Comment by jonas-schievink** [comment]
Ping from triage @mdnight, can you address the review comment?

**Comment by JohnCSimon** [comment]
Ping from triage @mdnight  please address the review comment

**Comment by JohnCSimon** [comment]
@scottmcm  Triage: Thanks for the PR but I'm going to close this due to lack of movement.

**Pull request closed without merging**